### PR TITLE
Adding PHP Backend

### PR DESCRIPTION
**Pull request by chwoda1**
Adding PHP backend for
parsing composer.json and
composer.lock files. Also
integrating with the packagist
API

Testin
* unit tests
* validation with commands in README

### CODE REVIEW

File: internal/backends/php/php.go [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+	packages := make(map[api.PkgName]api.PkgVersion)
+
+	for _, pkg := range composerLock.Packages {
+		packages[api.PkgName(pkg.Name)] = api.PkgVersion(pkg.Version)
+	}
+
+	return packages
+}
+
+var PhpComposerBackend = api.LanguageBackend{
```
Comment (masad-frost) [inline internal/backends/php/php.go:210]: Can we add Quirks `QuirksAddRemoveAlsoLocks` `QuirksAddRemoveAlsoInstalls`?

**Review by masad-frost — APPROVED** [review]
Other than the quirks this looks good to me!

@ryantm can you give it another pass and merge it?

**Review by ryantm — APPROVED** [review]
The code looks good to me and I tried all the Core and Index features and they seem to work. Thank you!

**Comment by ryantm** [comment]
@chwoda1 Looks like the TestInfoParsing test is failing currently:
```
$  go test -mod=readonly ./internal/backends/php/... -test.v
=== RUN   TestPackagistInfo
--- PASS: TestPackagistInfo (0.24s)
=== RUN   TestPackagistSearch
--- PASS: TestPackagistSearch (0.23s)
=== RUN   TestListSpecFile
--- PASS: TestListSpecFile (0.00s)
=== RUN   TestListLockFile
--- PASS: TestListLockFile (0.00s)
=== RUN   TestSearchParsing
--- PASS: TestSearchParsing (0.00s)
=== RUN   TestInfoParsing
Error parsing packagist payload
FAIL  github.com/replit/upm/internal/backends/php 0.482s
FAIL
```

**Comment by chwoda1** [comment]
> @chwoda1 Looks like the TestInfoParsing test is failing currently:
> 
> ```
> $  go test -mod=readonly ./internal/backends/php/... -test.v
> === RUN   TestPackagistInfo
> --- PASS: TestPackagistInfo (0.24s)
> === RUN   TestPackagistSearch
> --- PASS: TestPackagistSearch (0.23s)
> === RUN   TestListSpecFile
> --- PASS: TestListSpecFile (0.00s)
> === RUN   TestListLockFile
> --- PASS: TestListLockFile (0.00s)
> === RUN   TestSearchParsing
> --- PASS: TestSearchParsing (0.00s)
> === RUN   TestInfoParsing
> Error parsing packagist payload
> FAIL  github.com/replit/upm/internal/backends/php 0.482s
> FAIL
> ```

hm ok I'll post a new revision in a few hours with a fix + quirks change

Another thing: if you don't pass in a valid package-name `[vendor]/[packageName]` in on the info API it responds with a 200 but the data isn't JSON, causing some strange behavior. I'll add a regex check + a test-case for malformed json in the new revision

**Comment by chwoda1** [comment]
refactored the code a bit to avoid using `util.Die` so we could get better test coverage. Added the quirk stuff as well

**Review by ryantm — APPROVED** [review]
It all looks good now. Thank you for fixing it up!

**Pull request merged**